### ISSUE TITLE
Support for either pluggable HTTP clients or passing in Hackney options

### ISSUE THREAD

**Issue by sparkertime** [elixir-waffle/waffle#75]
Right now, there's no way to pass additional options through to hackney that I can find. This means I cannot use a proxy in my HTTP requests. That's required in my environment so Waffle is not usable at the moment.

ExAws gets around this in two ways:

- first, it lets you configure the [HttpClient](https://github.com/ex-aws/ex_aws/blob/master/lib/ex_aws/request/http_client.ex) through a defined behavior.
- second, if you choose to use hackney, it supports `hackey_opts` which get included in each request

I'd love to see the former but the latter would probably be simpler to implement and would also help.

**Comment by achempion** [comment]
Hi,

Waffle uses `:hackney` only to implement remote downloads and you can [configure the headers](https://hexdocs.pm/waffle/1.1.4/Waffle.Definition.Storage.html#module-passing-custom-headers-when-downloading-from-remote-path).

Could you, please, provide more context for your issue?

We use ExAws and you can [override the default configuration](https://hexdocs.pm/waffle/1.1.4/Waffle.Storage.S3.html).

**Comment by sparkertime** [comment]
Sure. In order to use an HTTP proxy, you need to pass in Hackney options (not headers) [as described here in the Hackney docs](https://github.com/benoitc/hackney#proxy-a-connection).

While I can override ExAws to use a proxy, I cannot do that for Waffle itself since I can only configure the headers, not the Hackney options. Since remote downloads fail in my environment without using an outbound HTTP proxy, that means I can't use Waffle even though ExAws works fine.

**Comment by sparkertime** [comment]
You can see an example of how ExAws supports this in Hackney through `hackney_opts` here - https://github.com/ex-aws/ex_aws/blob/6e6cb2655e74ac0fe88dea55257235efe90152b9/lib/ex_aws/request/hackney.ex#L18

**Comment by achempion** [comment]
The ExAws configuration does not depend on Waffle. You can configure it to your liking.

Just add this section to your configuration file

```elixir
config :ex_aws,
  json_codec: Jason,
  access_key_id: [{:system, "AWS_ACCESS_KEY_ID"}, :instance_role],
  secret_access_key: [{:system, "AWS_SECRET_ACCESS_KEY"}, :instance_role],
  hackney_opts: ... # YOUR CONFIGURATION IS HERE
```

**Comment by achempion** [comment]
Just to make it clear, what exactly do you want to configure?

Options to access custom S3 compatible endpoint or options to fetch images by url?

**Comment by achempion** [comment]
Ok, I get it, you want to use custom options for remote downloads.

Good point, we definitely need that. As a workaround you could download the image upfront and then pass it as binary/path to the Waffle.

**Comment by sparkertime** [comment]
> Ok, I get it, you want to use custom options for remote downloads.

Yup, that's correct. I think the best option would be making the HTTP backend for Waffle configurable like ExAws - it was pretty painless to write my own which used my proxied Finch setup instead of Hackney.

However, that's probably more work. Just providing a `:hackney_opts` option would be a great solution too.

**Comment by achempion** [comment]
Yes, indeed.

It should be relatively straight forward to implement, similar to `remote_file_headers/1` method.

We could introduce something like `remote_file_options/1` to merge it with the current options list [here](https://github.com/elixir-waffle/waffle/blob/master/lib/waffle/file.ex#L171).

Probably, better to allow per-uploader configuration with dynamic method call for more flexibility instead of static option in configuration file.
